docs/Encapsulation.md

```diff
@@ -0,0 +1,192 @@
+the second route (`/two`) does not, and the third route (`/three`) has
+access to the same context as the second route. Using
+[fastify-bearer-auth][bearer] to provide the authentication, the code for this
+example is as follows:
+
+```js
+'use strict'
+
+const fastify = require('fastify')()
+
+fastify.decorateRequest('answer', 42)
+
+fastify.register(async function authenticatedContext (childServer) {
+  childServer.register(require('fastify-bearer-auth'), { keys: ['abc123'] })
+
+  childServer.route({
+    path: '/one',
+    method: 'GET',
+    handler (request, response) {
+      response.send({
+        answer: request.answer,
+        foo: request.foo,
+        bar: request.bar
+      })
+    }
+  })
+})
+
+fastify.register(async function publicContext (childServer) {
+  childServer.decorateRequest('foo', 'foo')
+
+  childServer.route({
+    path: '/two',
+    method: 'GET',
+    handler (request, response) {
+      response.send({
+        answer: request.answer,
+        foo: request.foo,
+        bar: request.bar
+      })
+    }
+  })
+
+  childServer.register(async function grandchildContext (grandchildServer) {
+    grandchildServer.decorateRequest('bar', 'bar')
+
+    grandchildServer.route({
+      path: '/three',
+      method: 'GET',
+      handler (request, response) {
+        response.send({
+          answer: request.answer,
+          foo: request.foo,
+          bar: request.bar
+        })
+      }
+    })
+  })
+})
+
+fastify.listen(8000)
+```
+
+The above server example shows all of the encapsulation concepts outlined in the
+original diagram:
+
+1. Each _child context_, `authenticatedContext`, `publicContext`, and
+`grandchildContext` have access to the `answer` request decorator defined in
+the _root context.
+2. Only the `authenticatedContext` has access to the `fastify-bearer-auth`
+plugin.
+3. Both the `publicContext` and `grandchildContext` have access to the `foo`
+request decorator.
+4. Only the `grandchildContext` has access to the `bar` request decorator.
+
+To see this, start the server and issue requests:
+
+```sh
+# curl -H 'authorization: Bearer abc123' http://127.0.0.1:8000/one
+{"answer":42}
+# curl http://127.0.0.1:8000/two
+{"answer":42,"foo":"foo"}
+# curl http://127.0.0.1:8000/three
+{"answer":42,"foo":"foo","bar":"bar"}
+```
+
+[bearer]: https://github.com/fastify/fastify-bearer-auth
+
+<a id="shared-context"></a>
+## Sharing Between Contexts
+
+Notice that each context in the prior example inherits _only_ from the parent
+contexts. Sibibling contexts do not share any entities. This default is
+occassionally not desired. In such cases, the encapsulation context can be
+broken through the usage of [fastify-plugin][fastify-plugin] such that
+anything registered in a descendent context is availble to the containing
+parent context.
+
+Assuming the `publicContext` needs access to the `bar` decorator defined
+within the `grandchildContext` in the previous example, the code can be
+refactored to:
```

Comment:
"Refactoring" implies zero change in the observable behaviour, which I'm not sure in this case.

Reply:
```suggestion
rewritten to:
```

Reply:
Changed to "rewritten" (as I originally typed but, uh, refactored 🤣).